# Ex3 - Obtener y conocer sus datos

### Paso 1. Importar las bibliotecas necesarias

In [2]:
import pandas as pd

### Paso 2. Importe el conjunto de datos desde esta [dirección](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.user).

### Paso 3. Asígnalo a una variable llamada users y usa el 'user_id' como índice

In [3]:
users = pd.read_csv('https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.user', sep='|')
users.head()


user_id  age gender  occupation zip_code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

In [7]:
users = users.set_index('user_id') 
print(users.head())
print(users.index)


KeyError: "None of ['user_id'] are in the columns"

### Paso 4. Ver las 25 primeras entradas
Resuélvelo de 2 maneras diferentes

In [5]:
users[0:26]

age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
6         42      M      executive    98101
7         57      M  administrator    91344
8         36      M  administrator    05201
9         29      M        student    01002
10        53      M         lawyer    90703
11        39      F          other    30329
12        28      F          other    06405
13        47      M       educator    29206
14        45      M      scientist    55106
15        49      F       educator    97301
16        21      M  entertainment    10309
17        30      M     programmer    06355
18        35      F          other    37212
19        40      M      librarian    02138
20        42      F      homemaker    95660
21        26      M         writer    30068
22        25      M         writer    40206
23        30      F         artist    48197
24        21      F         artist    94533
25        39      M       engineer    55107
26        49      M       engineer    21044

In [6]:
users.iloc[0:26]

age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
6         42      M      executive    98101
7         57      M  administrator    91344
8         36      M  administrator    05201
9         29      M        student    01002
10        53      M         lawyer    90703
11        39      F          other    30329
12        28      F          other    06405
13        47      M       educator    29206
14        45      M      scientist    55106
15        49      F       educator    97301
16        21      M  entertainment    10309
17        30      M     programmer    06355
18        35      F          other    37212
19        40      M      librarian    02138
20        42      F      homemaker    95660
21        26      M         writer    30068
22        25      M         writer    40206
23        30      F         artist    48197
24        21      F         artist    94533
25        39      M       engineer    55107
26        49      M       engineer    21044

In [7]:
users.loc[0:25]

age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
6         42      M      executive    98101
7         57      M  administrator    91344
8         36      M  administrator    05201
9         29      M        student    01002
10        53      M         lawyer    90703
11        39      F          other    30329
12        28      F          other    06405
13        47      M       educator    29206
14        45      M      scientist    55106
15        49      F       educator    97301
16        21      M  entertainment    10309
17        30      M     programmer    06355
18        35      F          other    37212
19        40      M      librarian    02138
20        42      F      homemaker    95660
21        26      M         writer    30068
22        25      M         writer    40206
23        30      F         artist    48197
24        21      F         artist    94533
25        39      M       engineer    55107

### Paso 5. Ver las 10 últimas entradas

In [8]:
users.index

Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
       ...
       934, 935, 936, 937, 938, 939, 940, 941, 942, 943],
      dtype='int64', name='user_id', length=943)

In [9]:
users.loc[933:943]

age gender     occupation zip_code
user_id                                    
933       28      M        student    48105
934       61      M       engineer    22902
935       42      M         doctor    66221
936       24      M          other    32789
937       48      M       educator    98072
938       38      F     technician    55038
939       26      F        student    33319
940       32      M  administrator    02215
941       20      M        student    97229
942       48      F      librarian    78209
943       22      M        student    77841

In [8]:
users.tail(10)

age gender     occupation zip_code
user_id                                    
934       61      M       engineer    22902
935       42      M         doctor    66221
936       24      M          other    32789
937       48      M       educator    98072
938       38      F     technician    55038
939       26      F        student    33319
940       32      M  administrator    02215
941       20      M        student    97229
942       48      F      librarian    78209
943       22      M        student    77841

In [9]:
users.loc[len(users)-9:]

age gender     occupation zip_code
user_id                                    
934       61      M       engineer    22902
935       42      M         doctor    66221
936       24      M          other    32789
937       48      M       educator    98072
938       38      F     technician    55038
939       26      F        student    33319
940       32      M  administrator    02215
941       20      M        student    97229
942       48      F      librarian    78209
943       22      M        student    77841

### Paso 6. ¿Cuál es el número de observaciones del conjunto de datos?
Resuélvelo de 2 formas diferentes

In [10]:
#numeros de filas

In [10]:
len(users)

943

In [11]:
users.shape[0]

943

### Paso 7. ¿Cuál es el número de columnas del conjunto de datos?
Resuélvelo de 2 formas diferentes

In [12]:
users.shape[1]

4

In [14]:
len(users.columns)

4

In [12]:
users.info()

<class 'pandas.core.frame.DataFrame'>
Index: 943 entries, 1 to 943
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   age         943 non-null    int64 
 1   gender      943 non-null    object
 2   occupation  943 non-null    object
 3   zip_code    943 non-null    object
dtypes: int64(1), object(3)
memory usage: 69.1+ KB


### Paso 8. Imprime el nombre de todas las columnas.

In [13]:
print(users.columns)

Index(['age', 'gender', 'occupation', 'zip_code'], dtype='object')


In [14]:
users.keys()

Index(['age', 'gender', 'occupation', 'zip_code'], dtype='object')

### Paso 9. ¿Cómo se indexa el conjunto de datos? Obtener los valores del índice

In [16]:
users.index

Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
       ...
       934, 935, 936, 937, 938, 939, 940, 941, 942, 943],
      dtype='int64', name='user_id', length=943)

In [15]:
#utilizar el sciling o loc paraobtener el indice explicito
users.index[0:26]
users.loc[0:25]


age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
6         42      M      executive    98101
7         57      M  administrator    91344
8         36      M  administrator    05201
9         29      M        student    01002
10        53      M         lawyer    90703
11        39      F          other    30329
12        28      F          other    06405
13        47      M       educator    29206
14        45      M      scientist    55106
15        49      F       educator    97301
16        21      M  entertainment    10309
17        30      M     programmer    06355
18        35      F          other    37212
19        40      M      librarian    02138
20        42      F      homemaker    95660
21        26      M         writer    30068
22        25      M         writer    40206
23        30      F         artist    48197
24        21      F         artist    94533
25        39      M       engineer    55107

### Paso 10. ¿Cuál es el tipo de datos de cada columna?
Resuélvelo de 2 formas diferentes

In [16]:
users.dtypes

age            int64
gender        object
occupation    object
zip_code      object
dtype: object

In [17]:
users.info()

<class 'pandas.core.frame.DataFrame'>
Index: 943 entries, 1 to 943
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   age         943 non-null    int64 
 1   gender      943 non-null    object
 2   occupation  943 non-null    object
 3   zip_code    943 non-null    object
dtypes: int64(1), object(3)
memory usage: 69.1+ KB


In [17]:

print(users['age'].dtype)
print(users['gender'].dtype)
print(users['occupation'].dtype)
print(users['zip_code'].dtype)


int64
object
object
object


### Paso 11. Imprime sólo la columna de ocupación
Resuélvelo de 2 formas diferentes

In [19]:
users['occupation']

user_id
1         technician
2              other
3             writer
4         technician
5              other
           ...      
939          student
940    administrator
941          student
942        librarian
943          student
Name: occupation, Length: 943, dtype: object

In [20]:
users.occupation

user_id
1         technician
2              other
3             writer
4         technician
5              other
           ...      
939          student
940    administrator
941          student
942        librarian
943          student
Name: occupation, Length: 943, dtype: object

### Paso 12. ¿Cuántas ocupaciones diferentes hay en este conjunto de datos? ¿Cuántos usuarios por ocupación hay en el conjunto de datos?

In [18]:
#ocupaciones distintas
len(users['occupation'])

943

In [19]:
#ocupacion y usuarios
users['occupation'].value_counts()

occupation
student          196
other            105
educator          95
administrator     79
engineer          67
programmer        66
librarian         51
writer            45
executive         32
scientist         31
artist            28
technician        27
marketing         26
entertainment     18
healthcare        16
retired           14
lawyer            12
salesman          12
none               9
homemaker          7
doctor             7
Name: count, dtype: int64

In [18]:
users['occupation'].unique() #si pones nunique devuelve el numero 

array(['technician', 'other', 'writer', 'executive', 'administrator',
       'student', 'lawyer', 'educator', 'scientist', 'entertainment',
       'programmer', 'librarian', 'homemaker', 'artist', 'engineer',
       'marketing', 'none', 'healthcare', 'retired', 'salesman', 'doctor'],
      dtype=object)

In [20]:
users.groupby('occupation').sum()['age']

occupation
administrator    3061
artist            879
doctor            305
educator         3991
engineer         2438
entertainment     526
executive        1239
healthcare        665
homemaker         228
lawyer            441
librarian        2040
marketing         978
none              239
other            3625
programmer       2186
retired           883
salesman          428
scientist        1102
student          4328
technician        895
writer           1634
Name: age, dtype: int64

### Paso 13. ¿Cuál es la ocupación más frecuente?

In [37]:
users['occupation'].mode()

0    student
Name: occupation, dtype: object

con el metodo value counts se puede observar que la ocupacion más frecuente es estudiante con 196, tambien nos lo indica cuando sacamos la moda, ya que representa el valor que mas se repite dentro de la categoria ocupación

### Paso 14. Resuma el DataFrame.

In [20]:
users.info()

<class 'pandas.core.frame.DataFrame'>
Index: 943 entries, 1 to 943
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   age         943 non-null    int64 
 1   gender      943 non-null    object
 2   occupation  943 non-null    object
 3   zip_code    943 non-null    object
dtypes: int64(1), object(3)
memory usage: 69.1+ KB


### Paso 15. Resuma todas las columnas

In [21]:
users.describe()

age
count  943.000000
mean    34.051962
std     12.192740
min      7.000000
25%     25.000000
50%     31.000000
75%     43.000000
max     73.000000

In [22]:
users.describe(include=all)

TypeError: Cannot interpret '<built-in function all>' as a data type

### Paso 16. Resuma sólo la columna de ocupación

In [40]:
users.occupation

user_id
1         technician
2              other
3             writer
4         technician
5              other
           ...      
939          student
940    administrator
941          student
942        librarian
943          student
Name: occupation, Length: 943, dtype: object

### Paso 17. ¿Cuál es la edad media de los usuarios?

In [25]:
print(users['age'].describe())
print('----')
print('la media es de 34,05 años')

count    943.000000
mean      34.051962
std       12.192740
min        7.000000
25%       25.000000
50%       31.000000
75%       43.000000
max       73.000000
Name: age, dtype: float64
----
la media es de 34,05 años


### Paso 18. ¿Cuál es la edad con menos ocurrencias?

In [46]:
edad= users['age'].value_counts()
print(edad.to_dict())
print('La edad con menos ocurrencias es 7,11,66,10,73')

{30: 39, 25: 38, 22: 37, 28: 36, 27: 35, 26: 34, 24: 33, 20: 32, 29: 32, 23: 28, 32: 28, 21: 27, 35: 27, 33: 26, 31: 25, 44: 23, 19: 23, 39: 22, 36: 21, 42: 21, 40: 21, 50: 20, 51: 20, 48: 20, 49: 19, 37: 19, 18: 18, 34: 17, 38: 17, 45: 15, 47: 14, 17: 14, 43: 13, 46: 12, 53: 12, 55: 11, 41: 10, 60: 9, 57: 9, 56: 6, 15: 6, 52: 6, 16: 5, 13: 5, 54: 4, 65: 3, 61: 3, 59: 3, 14: 3, 58: 3, 70: 3, 63: 3, 69: 2, 62: 2, 64: 2, 68: 2, 7: 1, 11: 1, 66: 1, 10: 1, 73: 1}
La edad con menos ocurrencias es 7,11,66,10,73


In [26]:
users['age'].value_counts(ascending=True)

age
7      1
11     1
73     1
10     1
66     1
      ..
27    35
28    36
22    37
25    38
30    39
Name: count, Length: 61, dtype: int64